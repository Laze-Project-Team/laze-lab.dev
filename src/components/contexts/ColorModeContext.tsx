import type { PaletteMode, Theme } from '@mui/material';
import { CssBaseline, ThemeProvider, useMediaQuery } from '@mui/material';
import type { FC, ReactNode } from 'react';
import { createContext, useContext, useEffect, useMemo, useState } from 'react';

import { getTheme } from '@/lib/theme';

export type colorModeContextType = {
  toggleColorMode: () => void;
  colorMode: PaletteMode;
  theme: Theme;
};

const colorModeContext = createContext<colorModeContextType | null>(null);

export const useColorMode = (): colorModeContextType => {
  const colorMode = useContext(colorModeContext);

  if (colorMode === null) {
    throw new Error('Please wrap component with ColorModeProvider');
  }

  return colorMode;
};

export type colorModeProviderProps = {
  children?: ReactNode;
  preferTheme?: PaletteMode;
  isLocal?: boolean;
};

export const ColorModeProvider: FC<colorModeProviderProps> = ({
  children,
  preferTheme,
  isLocal,
}) => {
  const prefersDarkMode = useMediaQuery('(prefers-color-scheme: dark)');
  const [mode, setMode] = useState<PaletteMode>(
    preferTheme ?? prefersDarkMode ? 'dark' : 'light',
  );

  useEffect(() => {
    if (preferTheme !== undefined) {
      setMode(preferTheme);
    }
  }, [preferTheme]);

  // Update the theme only if the mode changes
  const theme = useMemo(() => getTheme(mode), [mode]);

  const colorMode = useMemo(
    () => ({
      // The dark mode switch would invoke this method
      toggleColorMode: () => {
        setMode((prevMode) => (prevMode === 'light' ? 'dark' : 'light'));
      },
      colorMode: mode,
      theme,
    }),
    [mode, theme],
  );

  return (
    <colorModeContext.Provider value={colorMode}>
      <ThemeProvider theme={theme}>
        {!isLocal && <CssBaseline />}
        {children}
      </ThemeProvider>
    </colorModeContext.Provider>
  );
};
