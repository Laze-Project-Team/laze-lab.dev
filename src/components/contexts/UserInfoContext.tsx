import type { FC, ReactNode } from 'react';
import { createContext, useContext } from 'react';

import type { userInfo } from '@/components/hooks/useUserInfo';
import { useUserInfoStrictly } from '@/components/hooks/useUserInfo';

export const userInfoContext = createContext<userInfo | null>(null);

export const useUserInfoContext = (): userInfo => {
  const userInfo = useContext(userInfoContext);

  if (userInfo === null) {
    throw new Error('please wrap component with UserInfoProvider');
  }

  return userInfo;
};

export const UserInfoProvider: FC<{
  children: ReactNode;
}> = ({ children }) => {
  const userInfo = useUserInfoStrictly();

  return (
    <userInfoContext.Provider value={userInfo}>
      {children}
    </userInfoContext.Provider>
  );
};
