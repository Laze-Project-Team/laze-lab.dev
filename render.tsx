import type {
  Queries,
  queries,
  RenderOptions,
  RenderResult,
} from '@testing-library/react';
import { render as defaultRender } from '@testing-library/react';
import type { FC, ReactNode } from 'react';

import { ColorModeProvider } from '@/components/contexts/ColorModeContext';

export type renderType<
  Q extends Queries = typeof queries,
  Container extends Element | DocumentFragment = HTMLElement,
  BaseElement extends Element | DocumentFragment = Container,
> = (
  ui: React.ReactElement,
  options?: RenderOptions<Q, Container, BaseElement>,
) => RenderResult<Q, Container, BaseElement>;

type wrapperProps = {
  children?: ReactNode;
};

type globalWrapperHOCType = (
  wrapper?: React.JSXElementConstructor<{ children: React.ReactElement }>,
) => FC<wrapperProps>;

const globalWrapperHOC: globalWrapperHOCType = (Wrapper) => {
  const globalWrapper: FC<wrapperProps> = ({ children }) => (
    <ColorModeProvider>
      {Wrapper ? (
        <Wrapper>
          <>{children}</>
        </Wrapper>
      ) : (
        children
      )}
    </ColorModeProvider>
  );

  return globalWrapper;
};

export const render: renderType = (ui, options) =>
  defaultRender(ui, {
    ...options,
    wrapper: globalWrapperHOC(options?.wrapper),
  });
