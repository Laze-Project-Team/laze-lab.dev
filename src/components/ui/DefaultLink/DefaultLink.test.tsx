import '@testing-library/jest-dom';

import { screen } from '@testing-library/react';
import { expect, it } from 'vitest';

import { render } from '/render';
import { ColorModeProvider } from '@/components/contexts/ColorModeContext';

import { DefaultLink } from '.';

it('should be rendered correctly', () => {
  const { container } = render(<DefaultLink href="/" />, {
    wrapper: ColorModeProvider,
  });
  expect(container.firstChild).toMatchSnapshot();
});

it('should have "link" role', () => {
  render(<DefaultLink href="/" />, { wrapper: ColorModeProvider });
  expect(screen.getByRole('link')).toBeInTheDocument();
});
