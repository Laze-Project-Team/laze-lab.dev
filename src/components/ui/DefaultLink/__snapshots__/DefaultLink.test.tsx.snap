// Vitest Snapshot v1

exports[`should be rendered correctly 1`] = `
<a
  class="css-1ln0tcx-PresentialStyledLink-PresentialStyledLink-PresentialDefaultLink-PresentialDefaultLink"
  href="/"
/>
`;
